src/crypto.ts

```diff
@@ -0,0 +1,427 @@
+	const signPromise = new Promise<SignFunction>(function (resolve, reject) {
+		getProvider().then(function (provider) {
+			realSign = provider.getSign(algorithm);
+			resolve(realSign);
+		}).catch(function (error) {
+			reject(error);
+		});
+	});
+
+	// Return a wrapper that will defer calls to the sign function until a provider has been loaded.
+	const signFunction = <SignFunction> function (key: Key, data: Data, codec?: Codec): Promise<ByteBuffer> {
+		if (realSign) {
+			return realSign(key, <any> data, codec);
+		}
+		return signPromise.then<ByteBuffer>(function (sign) {
+			return sign(key, <any> data, codec);
+		});
+	};
+
+	// Return a wrapper class that will defer calls until a provider has been loaded and an actual SignFunction instance
+	// has been created.
+	signFunction.create = function<T extends Data> (key: Key, codec?: Codec): Signer<T> {
+		return new SignerWrapper(signPromise, key, codec);
+	};
+
+	return signFunction;
+}
+
+/**
+ * Returns a promise that resolves to the current provider object.
+ */
+function getProvider(): Promise<CryptoProvider> {
+	if (provider) {
+		return Promise.resolve(provider);
+	}
+
+	if (providerPromise) {
+		return providerPromise;
+	}
+
+	providerPromise = new Promise(function (resolve, reject) {
+		// Load a platform-specific default provider.
+		if (typeof define === 'function' && define.amd) {
+			function loadProvider(mid: string) {
+				require([ mid ], function (_provider: CryptoProvider) {
+					// Don't overwrite a provider if one has already been set
+					if (!provider) {
+						provider = _provider;
+					}
+					resolve(provider);
+				});
+			}
+
+			if (has('host-node')) {
+				loadProvider('./providers/node');
+			}
+			else if (has('webcrypto')) {
+				loadProvider('./providers/webcrypto');
+			}
+			else {
+				loadProvider('./providers/script');
+			}
+		}
+		else if (has('host-node')) {
+			provider = require('./providers/node');
+			resolve(provider);
+		}
+		else {
+			reject(new Error('Unknown environment or loader'));
+		}
+	});
+
+	return providerPromise;
+}
+
+/**
+ * Sets the implementation provider.
+ *
+ * The provider may either be a loaded provider or a Promise that will resolve to a provider.
+ */
+export function setProvider(_provider: CryptoProvider | Promise<CryptoProvider>): void {
+	providerPromise = provider = null;
+
+	if (_provider) {
+		if ((<any> _provider).then) {
+			providerPromise = <Promise<CryptoProvider>> _provider;
+		}
+		else if ((<any> _provider).getHash) {
+			provider = <CryptoProvider> _provider;
+		}
+	}
+}
+
+/**
+ * An object for hashing a data stream.
+ */
+export interface Hasher<T extends Data> extends Sink<T> {
+	digest: Promise<ByteBuffer>;  // read only
+}
+
+/**
+ * Call a method that returns a promise, or return a resolved Promise if the method doesn't exist on the object
+ */
+function callOrNoop(object: any, methodName: string, ...args: any[]): Promise<any> {
+	const method: Function = object[methodName];
```

Comment:
should we check to make sure `object` is not `null` or `undefined` here before attempting to dereference a property off of it?